chapter07/dags/01_aws_handwritten_digits_classifier.py

```diff
@@ -0,0 +1,183 @@
+        "Id": "ExamplePolicy01",
+        "Statement": [
+            {
+                "Sid": "ExampleStatement01",
+                "Effect": "Allow",
+                "Principal": {
+                    "AWS": SAGEMAKER_ROLE
+                },
+                "Action": "s3:*",
+                "Resource": [
+                    f"arn:aws:s3:::{BUCKET_NAME}/*",
+                    f"arn:aws:s3:::{BUCKET_NAME}"
+                ]
+            }
+        ]
+    }
+
+    # Convert the policy from JSON dict to string
+    bucket_policy = json.dumps(bucket_policy)
+
+    # Set the new policy
+    s3hook = S3Hook()
+    session = S3Hook().get_session(region_name=REGION_NAME)
+    s3_client = session.client('s3')
+    s3_client.put_bucket_policy(Bucket=BUCKET_NAME, Policy=bucket_policy)
+
+def _extract_mnist_data():                          #B
+    s3hook = S3Hook()                               #C
+    # Download S3 dataset into memory
+    mnist_buffer = io.BytesIO()
+    mnist_obj = s3hook.get_key(                     #D
+        bucket_name=BUCKET_NAME,
+        key="mnist.pkl.gz",
+    )
+
+    mnist_obj.download_fileobj(mnist_buffer)
+    # Unpack gzip file, extract dataset, convert, upload back to S3
+    mnist_buffer.seek(0)
+
+    with gzip.GzipFile(fileobj=mnist_buffer, mode="rb") as f:
+        train_set, _, _ = pickle.loads(f.read(), encoding="latin1")
+        output_buffer = io.BytesIO()
+
+        write_numpy_to_dense_tensor(
+            file=output_buffer,
+            array=train_set[0],
+            labels=train_set[1],
+        )
+
+        output_buffer.seek(0)
+        s3hook.load_file_obj(                      #E
+            output_buffer,
+            key="mnist_data",
+            bucket_name=BUCKET_NAME,
+            replace=True,
+        )
+
+with DAG(
+    dag_id="01_aws_handwritten_digits_classifier",
+    schedule=None,
+    start_date=airflow.utils.dates.days_ago(3),
+):
+    create_bucket = S3CreateBucketOperator(
+        task_id="create_mnist_bucket",
+        bucket_name=BUCKET_NAME,
+        region_name=REGION_NAME
+    )
+
+    add_bucket_policy = PythonOperator(
+        task_id="add_sagemaker_bucket_policy",
+        python_callable=_add_bucket_policy,
+    )
+
+    download_mnist_data = S3CopyObjectOperator(    #A
+        task_id="download_mnist_data",
+        source_bucket_name="sagemaker-sample-data-eu-west-1",
+        source_bucket_key="algorithms/kmeans/mnist/mnist.pkl.gz",
+        dest_bucket_name=BUCKET_NAME,
+        dest_bucket_key="mnist.pkl.gz",
+    )
+
+    extract_mnist_data = PythonOperator(            #F
+        task_id="extract_mnist_data",
+        python_callable=_extract_mnist_data,
+    )
+
+    sagemaker_train_model = SageMakerTrainingOperator(            #G
+        task_id="sagemaker_train_model",
+        config={                                                  #H
+            "TrainingJobName": "mnistclassifier-{{ execution_date.strftime('%Y-%m-%d-%H-%M-%S') }}",
```

Comment:
nitpicking but i would add a global variable for 
DATE_FORMAT
%Y-%m-%d-%H-%M-%S

some char gains there, specially for the book layout